### PR TITLE
Improve project files discovery performance

### PR DESCRIPTION
**Pull request by dvz**
Partially resolves .#7057

### CODE REVIEW

**Comment by orklah** [comment]
Thanks!

File: src/Psalm/Internal/Analyzer/ProjectAnalyzer.php [hunk 1]
```diff
@@ -308,7 +308,11 @@ public function __construct(
...
         foreach ($this->config->getProjectDirectories() as $dir_name) {
-            $file_paths = $this->file_provider->getFilesInDir($dir_name, $file_extensions);
+            $file_paths = $this->file_provider->getFilesInDir(
+                $dir_name,
+                $file_extensions,
+                [$this->config, 'isInProjectDirs']
+            );
 
             foreach ($file_paths as $file_path) {
                 if ($this->config->isInProjectDirs($file_path)) {
```
Comment (weirdan) [inline src/Psalm/Internal/Analyzer/ProjectAnalyzer.php:318]: With the filter pushed down to the iterator, do we still need to check whether the returned file paths are in project / extra dirs?

Comment (dvz) [reply src/Psalm/Internal/Analyzer/ProjectAnalyzer.php:318]: Currently - yes (the iterator returns `true` for all leaves), but the filter can be made more universal: #7201

Comment (weirdan) [reply src/Psalm/Internal/Analyzer/ProjectAnalyzer.php:318]: All _first-level_ leaves, that is. But yeah, got you.

**Pull request merged**